### PR TITLE
System plugins - Adding new model into Storage layer

### PR DESCRIPTION
**Pull request by fernandofgfer**
#4604 - PR 1/2 - part from #4278 
## Description
This issue consists in add the new System plugin model to the Storage layer

## How
- Create new xcdatamodel version -> 54
- Add new entity to this datamodel
- Create NSManagedObject
- Add load and delete methods
- Update MIGRATIONS.md
- Add unit test for load an delete methods
- Add unit test for migration 53 -> 54
 
## How to test:
Launch unit tests

Update release notes:
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Storage/Storage.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -45,7 +47,7 @@
 		262B270E2621412000A421CF /* ProductAddOn+CoreDataProperties.swift in Sources */ = {isa = PBXBuildFile; fileRef = 262B26752621401A00A421CF /* ProductAddOn+CoreDataProperties.swift */; };
 		262B27132621412700A421CF /* ProductAddOnOption+CoreDataClass.swift in Sources */ = {isa = PBXBuildFile; fileRef = 262B26762621401A00A421CF /* ProductAddOnOption+CoreDataClass.swift */; };
 		262B27182621412D00A421CF /* ProductAddOnOption+CoreDataProperties.swift in Sources */ = {isa = PBXBuildFile; fileRef = 262B26772621401A00A421CF /* ProductAddOnOption+CoreDataProperties.swift */; };
-		263E37D72641AD5100260D3B /* BuildFile in Frameworks */ = {isa = PBXBuildFile; productRef = 263E37D62641AD5100260D3B /* SwiftPackageProductDependency */; };
+		263E37D72641AD5100260D3B /* Codegen in Frameworks */ = {isa = PBXBuildFile; productRef = 263E37D62641AD5100260D3B /* Codegen */; };
```
Comment (Ecarrion) [inline Storage/Storage.xcodeproj/project.pbxproj]: ⚠️  Are these changes necessary? 
Maybe we need to revert them.

File: Storage/Storage/Tools/StorageType+Extensions.swift [hunk 2]
```diff
@@ -537,4 +537,22 @@ public extension StorageType {
...
         return firstObject(ofType: PaymentGatewayAccount.self, matching: predicate)
-    }}
+    }
+
+    // MARK: - System plugins
+
+    /// Returns all stored system plugins for a provided `siteID`.
+    ///
+    func loadSystemPlugins(siteID: Int64) -> [SystemPlugin] {
+        let predicate = \SystemPlugin.siteID == siteID
```
Comment (Ecarrion) [inline Storage/Storage/Tools/StorageType+Extensions.swift:555]: Do you have any opinion on building keypaths like this?

File: Storage/Storage/Tools/StorageType+Extensions.swift [hunk 3]
```diff
@@ -537,4 +537,22 @@ public extension StorageType {
...
+        let descriptor = NSSortDescriptor(keyPath: \SystemPlugin.name, ascending: true)
```
Comment (Ecarrion) [inline Storage/Storage/Tools/StorageType+Extensions.swift:556]: [Q] Is there any reason to sort them by name?

File: Storage/Storage/Tools/StorageType+Deletions.swift [hunk 4]
```diff
@@ -129,4 +129,15 @@ public extension StorageType {
             deleteObject($0)
         }
     }
+
+    // MARK: - System plugins
```
Comment (Ecarrion) [inline Storage/Storage/Tools/StorageType+Deletions.swift]: Nit, I don't think it's necessary to add this MARKS, as we haven't been doing them in this file

File: Storage/Storage/Tools/StorageType+Deletions.swift [hunk 5]
```diff
@@ -129,4 +129,15 @@ public extension StorageType {
...
+
+    /// Deletes all of the stored SystemPlugins for the provided siteID.
+    ///
+    func deleteSystemPlugins(siteID: Int64) {
```
Comment (Ecarrion) [inline Storage/Storage/Tools/StorageType+Deletions.swift]: I think this function will be insufficient when you get to use it.
We try to only delete objects that are no longer necessary, if you look at the function above, you can see how we only delete stale entities.

⚠️ Do you think we can update this to follow that pattern?

**Review by Ecarrion — CHANGES_REQUESTED** [review]
Hi @fernandofgfer. This PR looks solid! I have a couple of requests marked with ⚠️ and questions.

Also, could you update the data model so properties are not marked as optional?

change


PS: I think you are missing some hound warnings too.

File: Storage/Storage/Tools/StorageType+Extensions.swift (hunk 2 above)
Comment (fernandofgfer) [reply Storage/Storage/Tools/StorageType+Extensions.swift:555]: I think is a well-readable way to do it! I like it

File: Storage/Storage/Tools/StorageType+Extensions.swift (hunk 3 above)
Comment (fernandofgfer) [reply Storage/Storage/Tools/StorageType+Extensions.swift:556]: Nope. I did it because maybe in the future we need to list it in the correct order, but I can remove it since we don't need it yet.

File: Storage/Storage.xcodeproj/project.pbxproj (hunk 1 above)
Comment (fernandofgfer) [reply Storage/Storage.xcodeproj/project.pbxproj]: I don't know why appeared these modifications, but I've reverted them.

File: Storage/Storage/Tools/StorageType+Deletions.swift (hunk 4 above)
Comment (fernandofgfer) [reply Storage/Storage/Tools/StorageType+Deletions.swift]: Ok, it's a personal way to separate concerts, but if you won't use it here, I will remove it.

File: Storage/Storage/Tools/StorageType+Deletions.swift (hunk 5 above)
Comment (fernandofgfer) [reply Storage/Storage/Tools/StorageType+Deletions.swift]: Sure! Modified!

**Comment by fernandofgfer** [comment]
Hi @Ecarrion!
I've updated the changes requested!

File: Storage/Storage/Tools/StorageType+Extensions.swift (hunk 3 above)
Comment (Ecarrion) [reply Storage/Storage/Tools/StorageType+Extensions.swift:556]: We can leave it, I was just curious about the intention. 
FYI we don't have to show this list anywhere for now...

**Review by Ecarrion — APPROVED** [review]
This looks great! 🎉 
I think you may need to merge develop to get this merged tho!
